File: Enzo.Music/Canzone/Parte.cs
using System.Numerics;

namespace Enzo.Music;

public class Parte : ICloneable, IEquatable<Parte>, IEqualityOperators<Parte, Parte, bool>
{
    public Accordo? Accordo;
    public string? Testo;

    public object Clone()
    {
        Parte p = new();
        if (Accordo != null) p.Accordo = Accordo.Clone() as Accordo;
        if (Testo != null) p.Testo = Testo;
        return p;
    }

    public bool Equals(Parte? other)
    {
        if (other == null) return false;
        bool uguali = String.Equals(Testo, other.Testo) && Accordo.Equals(Accordo, other.Accordo);
        return uguali;
    }
    public override bool Equals(object? obj)
    {
        if (obj is not Parte) return false;
        return Equals((Parte)obj);
    }

    public static bool Equals(Parte? a, Parte? b)
    {
        if (a == null || b == null) return false;
        return a.Equals(b);
    }

    public static new bool Equals(object? a, object? b)
    {
        if (a == null || b == null) return false;
        if (!(a is Parte && b is Parte)) return false;
        return ((Parte)a).Equals(b);
    }

    public override int GetHashCode()
    {
        return (Testo ?? string.Empty).GetHashCode();
    }

    public static bool operator ==(Parte? left, Parte? right)
    {
        return Equals(left, right);
    }

    public static bool operator !=(Parte? left, Parte? right)
    {
        return !Equals(left, right);
    }
}

File: KantaroBLL/Canzone/Parte.vb
Public Class Parte
    Implements ICloneable

    Public Property Accordo As Accordo
    Public Property Testo As String

    Public Function Clone() As Object Implements ICloneable.Clone
        Dim p As New Parte
        If Accordo IsNot Nothing Then p.Accordo = Accordo.Clone
        If Testo IsNot Nothing Then p.Testo = Testo
        Return p
    End Function

    Public Overrides Function Equals(obj As Object) As Boolean
        If obj Is Nothing Then Return False
        If Not TypeOf obj Is Parte Then Return False
        Dim b = DirectCast(obj, Parte)
        Dim uguali = String.Equals(Testo, b.Testo) AndAlso Accordo.Equals(Accordo, b.Accordo)
        Return uguali
    End Function

    Public Overloads Shared Function Equals(obj1 As Object, obj2 As Object) As Boolean
        If obj1 Is Nothing AndAlso obj2 Is Nothing Then Return True
        If Not TypeOf obj1 Is Parte Then Return False
        Return DirectCast(obj1, Parte).Equals(obj2)
    End Function

    Public Shared Operator =(a As Parte, b As Parte) As Boolean
        Return Parte.Equals(a, b)
    End Operator

    Public Shared Operator <>(a As Parte, b As Parte) As Boolean
        Return Not Parte.Equals(a, b)
    End Operator
End Class
